Markers #67

LDeeJay1969 opened the pull request:
For issue #67 markers are added: any shape can be marked as marker by right-clicking and selecting 'set as marker' from the context. From the menu on the left this name can then be selected and will become the centre of the screen. From this menu the marker can also be removed.

LDeeJay1969 commented:
Apologies for the messy commit. I forgot to update my locate repository so committed to an old dev at first.

Kruptein commented on server/api/socket/location.py:
```diff
@@ -51,7 +52,24 @@
+        .join(Layer)
+        .where((Layer.location == location))
+        .order_by(Initiative.index)
+    ]    
```
This should not be here anymore

LDeeJay1969 replied:
Ooops.... paste errror...

Kruptein commented on server/models/campaign.py:
```diff
@@ -166,6 +167,21 @@ def as_dict(self):
             self, recurse=False, exclude=[Note.room, Note.location, Note.user]
         )
 
+class Marker(BaseModel):
```
I see some issues here.  You declare both room and location.  This can only lead to potential bugs (e.g. a marker with room=a and location=l  where l is a location in room b).  You have access to room via the location refkey, so there is no need for a direct link to room.

Coincidentally there is actually also no real reason to have a direct link to location.  What I think you actually want, is to have the uuid be a reference to a shape directly, as a Marker in its current form is always associated with a particular shape.  Take for example ShapeLabel in shape.py, it has 2 direct foreign keys.  One to shape and one to label.  What we want here ultimately is 1 reference to shape and 1 to user, nothing more.

LDeeJay1969 replied:
I'm so annoyed (with myself) at the moment. I had the process working with strings instead of objects, then when hitting refresh, all markers were read as objects. So I changed the class to return a string instead of objects, now that works, now I can't write to the db. Whenever I want to, the server responds with "Task exception was never retrieved" on executing the `Marker.get_or_none(uuid=data)` where I'm sure data is a string (according to the log), not an object but when I change data to data.uuid, I get "'str' object has no attribute 'uuid'". That makes sense but I have no clue how to change the marker.py contents to read the db as string instead of object or to get rid of the warning.

LDeeJay1969 replied:
Nevermind...... just a contraint on the table that I completely overlooked. All's well again...

Kruptein commented on server/api/socket/location.py:
```diff
@@ -51,7 +52,24 @@
+        "Markers.Set",
+        [
+            marker.as_dict()
+            for marker in Marker.select().where(Marker.user == user)
```
This will return all the markers you own irregardless of the session/location you're in.

LDeeJay1969 replied:
And that's what you want, right?

LDeeJay1969 replied:
No you don't ..... I'll add them...

Kruptein replied:
There is still extra filtering required.  Imagine you have two completely different sessions in which you use markers.  This will return all markers from both sessions combined, which you wouldn't want.

LDeeJay1969 replied:
Now taken care of.

Kruptein commented on client/src/game/api/events.ts:
```diff
@@ -96,6 +96,9 @@ socket.on("Position.Set", (data: { x: number; y: number; zoom: number }) => {
 socket.on("Notes.Set", (notes: Note[]) => {
     for (const note of notes) gameStore.newNote({ note, sync: false });
 });
+socket.on("Markers.Set", (markers: Marker[]) => {
```
What you receive here is not in fact a Marker, but a Pick<Marker, 'uuid'> or rather just a string[].  The Marker.name aspect does not come from the server.

That said, I have some remarks on the Marker.name below

Kruptein commented on client/src/game/comm/types/general.ts:
```diff
@@ -67,3 +67,9 @@ export interface Note {
+
+export interface Marker {
+    uuid: string;
+    name: string;
```
I don't fully see the need for a separate name field here.

You have the uuid so you can always retrieve the shape name at any time.  If anything, storing the name in 2 different locations is bound for some sync issues.  I can already tell that at this point, renaming a shape will in fact not rename the marker until you refresh the page.

I think you want to keep the markers as a simple string[] array containing uuids in the current marker idea.

LDeeJay1969 replied:
This was kinda by design.... I thought that if you want to mark shapes with similar names you can differentiate them by naming them, marking them and then renaming them. You marker will then have a different, more distinctive name. But you're right, it doesn't have to be stored and might lead to confusion.

LDeeJay1969 replied:
On that note, if you want to make a more elaborate system where you want to 'share' makers etc, you might want to keep a name. And access-info (who has it), But realizing that, you might want to change the shape-focused marker to a name-to-shape focused marker so you can share the same shape with different marker-names....  tricky one. 

I'll remove it for now, it can always be added afterwards...

LDeeJay1969 replied:
And in doing so I just realized I wanted to keep them in an object to keep track of the two. In the menu on the left I display the name but keep track of the uuid. This way you can add markers with similar names but not with a duplicate uuid.

LDeeJay1969 replied:
This will be reset when you refresh though. If you want to keep them, you'll need to store them in the db as well.

Kruptein replied:
As you point out yourself, there is actually no mechanism to rename a marker and even if there was it would be overwritten on load by the shape name, unless you remove that aspect and set the default name for a new marker to the shape it is attached to.

You don't need to track the name in an object in order to display the name.  Just request the shape from the layermanager and show that name instead.


tl;dr  Either you offer true marker renaming where it is not associated with the shape itself or you only track the shape uuid as keeping track of the name on top of this is just bound to fail.

LDeeJay1969 replied:
That's what I don't get. Once it's pushed to array to show the name, I have no way of knowing from which shape-uuid it came unless I keep them together. Besides, once you click on the name, you want to know to which shape-uuid to go to. So you need to keep a relation between the two.

Kruptein replied:
`layerManager.UUIDMap.get(marker).name` will give you the name of the shape with uuid=marker

Kruptein commented on client/src/game/ui/menu/menu.vue:
```diff
@@ -68,6 +68,25 @@
+                                            </td>
+                                        </tr>
+                                    </tbody>
+                                </table>
```
I'm not super fond of <tables> :p If you can, rework it to a flexbox or grid, if you're not that familiar with it, I don't mind doing this as it shouldn't take a lot of effort

LDeeJay1969 replied:
Maybe not but this was twofold: firstly, when I used your approach, when making a lot of markers and notes, the menu bar gets a scroll bar. Scrolling would make the text and the icons move independently from each other (it still happens with the + for the notes). Using tables created a more 'sturdy' design which moved them together. It might have to do with the width that changes when adding the scrollbar but didn't want to mess with it too much.
Secondly it feel this gives the programmer a slightly more clear view about the design rather than having a lot of divs and having to figure out which class does what. But that's entirely personal.

Kruptein replied:
Using css grid is just simpler imo.  The 99% width and nowrap whitespace stuff feels wrong.

There is no absolute right/wrong here as this is just my preference ;)

LDeeJay1969 commented:
Changed from object to string, resolved the delete marker on delete shape and reworked the menu from table to flex. Still to do is set locations as marker. This can be added later though...

Kruptein commented on server/models/campaign.py:
```diff
@@ -166,6 +167,15 @@ def as_dict(self):
         )
 
+class Marker(BaseModel):
+    uuid = TextField(primary_key=True)
```
I would make the uuid a ForeignKeyField of Shape, as that's what it ultimately is.

Also, right now you'll get a rejection from the database if another user tries to set a marker on a shape someone else has already marked as uuid is a primary key.

LDeeJay1969 commented on client/src/game/shapes/utils.ts:
```diff
@@ -156,6 +156,7 @@ export function deleteShapes(): void {
         l.removeShape(sel, SyncMode.FULL_SYNC);
         EventBus.$emit("SelectionInfo.Shape.Set", null);
         EventBus.$emit("Initiative.Remove", sel.uuid);
+        gameStore.removeMarker({ marker: sel.uuid, sync: true }); //delete marker.... even if it doesn't exist.
```
Moved... 
Should there be a check to see whether the shape is marked or can I just try to remove the marker, even it might not exist? If I need to check first, I need to access (or create) the marker array which layer doesn't have direct access to, which means more references and arrays.

Kruptein commented:
Let me know when I can review this again.
Also some small advice, small commits are great, but try to push them in larger batches ;) I get a lot of emails currently :D

LDeeJay1969 commented:
Apologies for the commits, whilst trying to rearrange the code, I run into some issues....

I've checked for sessions now so you don't delete markers between different sessions. All works fine but I have to refresh the browser to see the menu. When I just switch sessions, the menu (cog-icon) won't work until I refresh the browser. When I remove `return layerManager.UUIDMap.get(marker)!.name;` from menu.vue it works ok, apart from showing [?] instead of the name (understandably). I can't figure out where this behaviour is originated.

Kruptein commented:
I'll try to give this a look in the weekend, so that we can hopefully get this working when I want to release 0.12.

I want to focus my efforts this week in one last ticket so that I can make a 0.11 release this week.

I haven't tested anything, but your `When I remove return layerManager.UUIDMap.get(marker)!.name;` sentence seems to imply to me that the returned value is not in fuck defined (the !),  Are there any console logs that can help you further until I take a more in depth look?

LDeeJay1969 commented:
I can switch between sessions now (the !  was indeed the issue). But the solution for this, poses the next problem: when loading the page, all is fine (all markers are named OK). When switching between sessions, _all_ markers are listed (also the ones from other sessions) and the ones not belonging to the current session are shown as [?].  So loading the page has the right query, switching sessions doesn't...

LDeeJay1969 commented:
Finally .... I figured out I didn't reset the list properly between switching sessions so once that was solved it works a treat! I have thoughts for adding markers on locations as well but I'll make a new issue for that one.

Kruptein commented:
Glad to hear you got it sorted out, I release 0.19 earlier , so I should have some more time now to properly review this PR again

LDeeJay1969 commented:
The only thing I haven't working yet is the marker as foreign key of Shape. Once I tried to change that, a whole lot of errors emerged on the server side and it wouldn't start anymore until recompiled so I have to dig deeper into that. Not sure if that's a huge issue at the moment though.

LDeeJay1969 commented:
Finally got the foreign key working. Probably not the most elegant(?) but it works. Since I had to switch to a as_dict (what Shape is) I struggled getting just a string back. But got a solution in the end. Let me know what you think.

Kruptein commented on server/models/campaign.py:
```diff
@@ -166,6 +167,20 @@ def as_dict(self):
+
+    uuid = ForeignKeyField(Shape, backref="markers", on_delete="CASCADE")
+    user = ForeignKeyField(User, backref="markers", on_delete="CASCADE")
+    room = ForeignKeyField(Room, backref="markers", on_delete="CASCADE")
```
Any particular reason why you chose to use Room and not Location?

LDeeJay1969 replied:
Nope. I started originally with a copy of Note because it looked similar to what I wanted. After a lot of messing about with a lot a files, the result is completely different :-) 
I've changed it to reference location instead of room.

Kruptein commented on server/save.py:
```diff
@@ -294,7 +294,9 @@ def upgrade(version):
-
+        db.execute_sql(
+            'CREATE TABLE IF NOT EXIST "marker" ("id" INTEGER NOT NULL PRIMARY KEY, `uuid` TEXT NOT NULL , `user_id` INTEGER NOT NULL, `room_id` INTEGER NOT NULL, FOREIGN KEY(`uuid`) REFERENCES `shape`(`uuid`) ON DELETE CASCADE, FOREIGN KEY(`room_id`) REFERENCES `room`(`id`) ON DELETE CASCADE, FOREIGN KEY(`user_id`) REFERENCES `user`(`id`) ON DELETE CASCADE)'
+        )
```
This needs to move to a separate elif clause

LDeeJay1969 replied:
As with changelog and menu.vue, I missed a few updates. Will redo this on the latest dev (probably my lack of experience with github as well ;-)

Kruptein replied:
This also seems to be still wrong

LDeeJay1969 replied:
Yes, as said for some reason I didn't see those new updates. Now updated.

LDeeJay1969 commented:
I'm so sorry for this mess.....  it's  very steep learning curve for me. 
But I think the penny finally dropped where I went wrong. Will update again.

LDeeJay1969 commented on client/src/game/store.ts:
```diff
@@ -246,6 +248,36 @@ class GameStore extends VuexModule implements GameState {
+    @Mutation
+    newMarker(data: { marker: string; sync: boolean }): void {
+        let mrk = "";
+        if (typeof data.marker === "object") {
```
Resolved this now.... turned back into a proper string. Went wrong on the server side.

LDeeJay1969 commented on client/src/game/store.ts:
```diff
@@ -246,6 +248,36 @@ class GameStore extends VuexModule implements GameState {
+        }
+        const exists = this.markers.some(m => m === mrk);
+        if (!exists) this.markers.push(mrk);
+        if (data.sync) socket.emit("Marker.New", mrk);
```
Again, went wrong at the server side. Resolved now.

LDeeJay1969 commented on client/src/game/store.ts:
```diff
@@ -246,6 +248,36 @@ class GameStore extends VuexModule implements GameState {
+
+    @Mutation
+    jumpToMarker(marker: string): void {
+        shape.uuid = layerManager.UUIDMap.get(marker);
```
Must have been a copy/paste error. Not what I had originally. Now changed.

LDeeJay1969 commented on client/src/game/ui/selection/shapecontext.vue:
```diff
@@ -77,6 +78,7 @@ import { Layer } from "../../layers/layer";
     computed: {
         ...mapState("game", ["activeFloorIndex", "assets", "locations", "notes"]),
         ...mapMutations("game", ["selectFloor"]),
+        ...mapState("game", ["locations", "assets", "notes", "markers"])
```
This occurred after a few updates. I'll recommit to the most current version (same goes for changelog).

Kruptein replied:
This seems to still be wrong?

LDeeJay1969 replied:
Should be solved now.

LDeeJay1969 commented on client/src/game/ui/selection/shapecontext.vue:
```diff
@@ -50,6 +50,7 @@
         <li @click="addInitiative">{{ getInitiativeWord() }} initiative</li>
         <li @click="deleteSelection">Delete shapes</li>
         <li v-if="hasSingleShape()" @click="openEditDialog">Show properties</li>
+        <li @click="setMarker">Set marker</li>
```
Good point. Didn't think of that. Changed now.

LDeeJay1969 commented:
Ok, think I'm done now (again). I see at the top of this page: 

> commits into Kruptein:dev from LDeeJay1969:feature/GH-67-markers

but aparently I'm not seeing the latest? Because save.py and changelog.md are still 'old' ones. How do I rectify that?

LDeeJay1969 commented on client/src/game/ui/menu/menu.vue:
```diff
@@ -196,6 +213,22 @@ export default class MenuBar extends Vue {
+        if (layerManager.UUIDMap.get(marker) != undefined) {
+            return layerManager.UUIDMap.get(marker)!.name;
+        } else {
+            return "";
```
This should not occur anymore since there's either a marker or there isn't. I can't think of a situation where you get a return but not know what. The undefined check is more to satisfy TS, I'm certain you won't get undefined back from the server.

Kruptein replied:
That's fine and good to hear that it shouldn't occur in normal situations.

I would still prefer the [?] to be returned here instead of in the template though.

Kruptein commented:
if you merge dev into your branch, did you resolve all the merge conflicts git throws at you ?

LDeeJay1969 commented:
I did now. I'm still confused about CHANGELOG.MD though. It wanted me to change lines that weren't mine....? I hope I got that right. I didn't get any conflicts about save.py and menu.vue. Not sure how that worked out then...

LDeeJay1969 commented on client/src/game/ui/menu/menu.vue:
```diff
@@ -196,6 +213,22 @@ export default class MenuBar extends Vue {
+
+    nameMarker(marker: string): string {
+        if (layerManager.UUIDMap.get(marker) != undefined) {
+            return layerManager.UUIDMap.get(marker)!.name;
```
Done that

Kruptein replied:
This is not what I meant :D 

```typescript
const shape = layerManager.UUIDMap.get(marker);
if (shape !== undefined) {
    return shape.name;
}
```
with the abvoe you only do the .get on UUIDmap once and guarantee type safeness inside the if check without an extra ! needed.

LDeeJay1969 replied:
I don't get it..... that's more or less what I had but it didn't show up on github...?

Kruptein commented on CHANGELOG.md:
```diff
@@ -43,6 +45,8 @@ All notable changes to this project will be documented in this file.
     -   Create/Delete floors that are rendered on top of eachother to increase immersion
     -   Use Page Up/Down as a quick keybinding to move between floors
+-  Touch Gestures [ZachMyers3]
+-  Easier client traversing by removing _load route [ZachMyers3] 
```
Yeah, something went wrong during your merge conflict :D

These lines are already there below, so just remove them again.

LDeeJay1969 replied:
Sorted.

Kruptein commented on client/src/game/store.ts:
```diff
@@ -244,6 +246,32 @@ class GameStore extends VuexModule implements GameState {
+
+    @Mutation
+    jumpToMarker(marker: string): void {
+        const shape = layerManager.UUIDMap.get(marker);
```
You should handle the case where the shape does not exist.

LDeeJay1969 replied:
Not sure what you mean here.... in which function?

LDeeJay1969 replied:
Got it. Updated now.

Kruptein commented on client/src/game/ui/menu/menu.vue:
```diff
@@ -223,8 +223,9 @@ export default class MenuBar extends Vue {
-            return layerManager.UUIDMap.get(marker)!.name;
+        const mrk = layerManager.UUIDMap.get(marker)!.name;
+        if (mrk !== undefined) {
+            return mrk;
```
```suggestion
        const shape = layerManager.UUIDMap.get(marker);
        if (shape !== undefined) {
            return shape.name;
```
I don't know why you didn't just use the suggestion I proposed. Your current approach will still error on that first line if the returned value is undefined.

LDeeJay1969 replied:
This was the only way I could get it working..... Your suggestion gives me an error: 
> Function lacks ending return statement and return type does not include 'unde
> fined'.

Kruptein replied:
It's only the start that needs to be replaced that I posted.
The entire function is

```typescript
nameMarker(marker: string): string {
        const shape = layerManager.UUIDMap.get(marker);
        if (shape !== undefined) {
            return shape.name;
        } else {
            return "";
        }
    }
```

LDeeJay1969 replied:
Probably my lack of TS-knowledge. I still don't get the difference but hey... it works! ;-)
I'll try to dig up some TS course. It's not really JS... ;-)

Kruptein replied:
your first line calls the UUIDMap.get function which can return either a value or undefined.  Then you call .name on that result.  If you do it straight away TS will warn you that this can result in an error if get returns undefined.  You then added a ! to tell TS that it does not need to worry and you're sure you know what you're doing.
Thing is that that is bound to go wrong at one point and thus better avoided.  Here avoiding it is relatively easy by calling .get() and then using an if check to verify whether it is undefined or not.

Kruptein commented on server/save.py:
```diff
@@ -311,6 +308,13 @@ def upgrade(version):
+    elif version == 21:
+        db.foreign_keys = False
+        db.execute_sql(
+            'CREATE TABLE IF NOT EXIST "marker" ("id" INTEGER NOT NULL PRIMARY KEY, `uuid` TEXT NOT NULL , `user_id` INTEGER NOT NULL, `room_id` INTEGER NOT NULL, FOREIGN KEY(`uuid`) REFERENCES `shape`(`uuid`) ON DELETE CASCADE, FOREIGN KEY(`room_id`) REFERENCES `room`(`id`) ON DELETE CASCADE, FOREIGN KEY(`user_id`) REFERENCES `user`(`id`) ON DELETE CASCADE)'
```
This SQL line still references room instead of location

LDeeJay1969 replied:
I hate these local/remote versions... ;-)
Copied from the wrong version. Apologies. Now corrected.

Kruptein commented on server/save.py:
```diff
@@ -309,6 +308,13 @@ def upgrade(version):
+    elif version == 21:
+        db.foreign_keys = False
+        db.execute_sql(
+            'CREATE TABLE IF NOT EXIST "marker" ("id" INTEGER NOT NULL PRIMARY KEY, `uuid` TEXT NOT NULL , `user_id` INTEGER NOT NULL, `location_id` INTEGER NOT NULL, FOREIGN KEY(`uuid`) REFERENCES `shape`(`uuid`) ON DELETE CASCADE, FOREIGN KEY ("location_id") REFERENCES "location" ("id") ON DELETE CASCADE, FOREIGN KEY(`user_id`) REFERENCES `user`(`id`) ON DELETE CASCADE)'
```
```suggestion
            'CREATE TABLE IF NOT EXISTS "marker" ("id" INTEGER NOT NULL PRIMARY KEY, `uuid_id` TEXT NOT NULL, `user_id` INTEGER NOT NULL, `location_id` INTEGER NOT NULL, FOREIGN KEY(`uuid_id`) REFERENCES `shape`(`uuid`) ON DELETE CASCADE, FOREIGN KEY ("location_id") REFERENCES "location" ("id") ON DELETE CASCADE, FOREIGN KEY(`user_id`) REFERENCES `user`(`id`) ON DELETE CASCADE)'
```

Kruptein replied:
There was a typo in the exists keyword and the uuid field was missing its _id postfix

LDeeJay1969 replied:
Changed

Kruptein commented on server/models/campaign.py:
```diff
@@ -166,6 +167,18 @@ def as_dict(self):
+    location = ForeignKeyField(Location, backref="markers", on_delete="CASCADE")
+
+    def __repr__(self):
+        return f"<Marker {self.shape.uuid} {self.location.get_path()} - {self.user.name}"
```
self.shape does not exist.  to fix it it should be self.uuid.uuid, which is kinda awkward.  So I suggest that you rename the uuid field to shape (don't forget to change the save.py migration query and marker.py socket file if you do this)

LDeeJay1969 replied:
Changed.

Kruptein commented on client/src/game/ui/menu/menu.vue:
```diff
@@ -324,14 +358,18 @@ DIRECTORY.CSS changes
     display: flex;
-    justify-content: space-evenly;
     align-items: center;
+    text-align: left;
```
I would apply this change only on the markers part for now.

Some redesign of the side bar is needed in the future, but having all the other elements also left-aligned is kinda weird imo.

(same thing applies to the removal of the justify-content)

LDeeJay1969 replied:
Done.

Kruptein commented on client/src/game/ui/selection/shapecontext.vue:
```diff
@@ -50,6 +50,7 @@
         <li @click="addInitiative">{{ getInitiativeWord() }} initiative</li>
         <li @click="deleteSelection">Delete shapes</li>
         <li v-if="hasSingleShape()" @click="openEditDialog">Show properties</li>
+        <li v-if="hasSingleShape()" @click="setMarker">Set marker</li>
```
After playing with it, we might want to hide this if it already is a marker or maybe replace it with a remove marker element.

LDeeJay1969 replied:
Right, I've been trying this. vue side is ok, gamestore side is ok. I can query the markers array in the gamestore from vue but I can't get the result back to vue, it's always undefined. Do I need to export something from gamestore or do I need to query the db directly from vue?

Kruptein replied:
You could add a method to the gamestore, but you do have everything you need in this file.  You import the markers from the store in your computed section and you have the access to the shape and thus it's id.

LDeeJay1969 commented:
Do you still want the marker removed from campaign.py?  I've changed the location so that Layer is accessed before Marker and that resolved the error (in my case).

Kruptein commented:
It's not a major show stopper, but I do in general like the imports to be sorted if possible, but the primary reason is that the campaign.py file is becoming pretty large.  So splitting it up wouldn't hurt I think.

Kruptein commented on server/api/socket/marker.py:
```diff
@@ -0,0 +1,36 @@
+    user = sid_data["user"]
+    location = sid_data["location"]
+
+    marker = Marker.get_or_none(uuid=uuid)
```
```suggestion
    marker = Marker.get_or_none(shape=uuid, user=user)
```
the `uuid` field no longer exists and additionally you need to explicitly check the user, or you could accidentally remove someone else's marker of the same shape.

LDeeJay1969 replied:
Done

Kruptein commented on server/api/socket/marker.py:
```diff
@@ -0,0 +1,36 @@
+    user = sid_data["user"]
+    location = sid_data["location"]
+
+    marker = Marker.get_or_none(shape=data)
```
```suggestion
    marker = Marker.get_or_none(shape=data, user=user)
```

LDeeJay1969 replied:
Changed

Kruptein commented on client/src/game/ui/menu/menu.vue:
```diff
@@ -68,6 +68,21 @@
                                 <div v-if="!notes.length">No notes</div>
                             </div>
                         </div>
+                        <!-- MARKERS -->
```
I would move the markers section out of the DM only section, I can see some players also wanting to jump around between markers in some circumstances.

LDeeJay1969 replied:
Done.

Kruptein commented:
I also changed something else that affects the save format, so you'll need to pull some changes and readjust your save.py file.  I know it's a bit clunky, so sorry for the inconvenience

LDeeJay1969 commented:
I hope I got the save.py concept right. I don't undestand though, why under version 20 Booleanfield is mentioned twice and under 21 Integerfield is mentioned as well (as is Booleanfield twice).

LDeeJay1969 commented:
What I also don't get is how my own gh-rep will keep up-2-date with dev. It's always conflicting with CHANGELOG, now also with save.py.  When I look at dev directly, I see completely different versions. I thought everything in dev would be reflected in my fork?

Kruptein commented:
> I hope I got the save.py concept right. I don't undestand though, why under version 20 Booleanfield is mentioned twice and under 21 Integerfield is mentioned as well (as is Booleanfield twice).

Haha that' actually just a stupid error on my end :p there is no need for it indeed.

Kruptein commented:
> What I also don't get is how my own gh-rep will keep up-2-date with dev. It's always conflicting with CHANGELOG, now also with save.py. When I look at dev directly, I see completely different versions. I thought everything in dev would be reflected in my fork?

Are you actively pulling in my dev into your dev ?

Kruptein commented on server/api/socket/marker.py:
```diff
@@ -0,0 +1,36 @@
+    user = sid_data["user"]
+    location = sid_data["location"]
+
+    marker = Marker.get_or_none(shape_id=uuid)
```
This one is not updated yet ?

LDeeJay1969 replied:
Yes, I did.... I noticed that `marker = Marker.get_or_none(uuid=uuid)` hadn't changed yet. Since the field name in the table is shape_id I changed it.

Kruptein replied:
No I mean it's missing user=user

LDeeJay1969 replied:
My bad, I changed the add-function instead of the delete-function.... sorted now.

Kruptein replied:
😄 It needs to be there in both cases

Kruptein replied:
otherwise other users won't be able to create a marker for a similar shape

LDeeJay1969 replied:
I know I shouldn't do two things at the same time.... 
Focus Leon, FOCUS!!!  ;-)

LDeeJay1969 commented:
Yay !! Apologies for the mess. It was a steep learning curve (still ongoing) for someone with no knowledge of TS and VUE but I've learned a lot so thanks for the input and the patience. 

I've got some general questions (PA related but not directly functional). Where do you want me to pose those?

Kruptein commented:
Hey no problems at all, I'm not gonna claim this to be the easiest or cleanest code base by any means, so some struggles are very normal.  If you happen to have discord, I created a server yesterday, https://discord.gg/mubGnTe.   Otherwise feel free to mail me or send me something on reddit :)
